### PR TITLE
fix #6899 handle Unicode output in activate commands

### LINKED ISSUE #6899: activate script fails with conda=4.4.10 and zsh=5.3.1 with cryptic logging error
**I'm submitting a...**
  - [x] bug report
  - [ ] feature request

## Current Behavior

Using ZSH 5.3.1 the activate script of `conda=4.4.10` does not work and fails with some cryptic Python logging error. It doesn't even detect an error. With 4.4.0 it at least fails at unicode-decoding `PS1` and offers to submit a report.
The script works in Bash.
### Steps to Reproduce

```
$ source ~/anaconda/bin/activate
Traceback (most recent call last):
  File "/home/ondrej/anaconda/lib/python2.7/site-packages/conda/gateways/logging.py", line 64, in emit
    msg = self.format(record)
  File "/home/ondrej/anaconda/lib/python2.7/logging/__init__.py", line 734, in format
    return fmt.format(record)
  File "/home/ondrej/anaconda/lib/python2.7/logging/__init__.py", line 465, in format
    record.message = record.getMessage()
  File "/home/ondrej/anaconda/lib/python2.7/logging/__init__.py", line 329, in getMessage
    msg = msg % self.args
  File "/home/ondrej/anaconda/lib/python2.7/site-packages/conda/__init__.py", line 43, in __repr__
    return '%s: %s' % (self.__class__.__name__, text_type(self))
  File "/home/ondrej/anaconda/lib/python2.7/site-packages/conda/__init__.py", line 47, in __str__
    return text_type(self.message % self._kwargs)
ValueError: unsupported format character '{' (0x7b) at index 458
Logged from file exceptions.py, line 770
$ echo $ZSH_VERSION 
5.3.1
```

## Expected Behavior

The source script should work as it did with `conda<=4.3*`. My current workaround is to prepend `~/anaconda/bin` to `$PATH` manually.

## Environment Information
<details open><summary><code>`conda info`</code></summary><p>

```

     active environment : base
    active env location : /home/ondrej/anaconda
            shell level : 1
       user config file : /home/ondrej/.condarc
 populated config files : /home/ondrej/.condarc
          conda version : 4.4.10
    conda-build version : 1.19.0
         python version : 2.7.13.final.0
       base environment : /home/ondrej/anaconda  (writable)
           channel URLs : https://conda.anaconda.org/anaconda-fusion/linux-64
                          https://conda.anaconda.org/anaconda-fusion/noarch
                          https://repo.continuum.io/pkgs/main/linux-64
                          https://repo.continuum.io/pkgs/main/noarch
                          https://repo.continuum.io/pkgs/free/linux-64
                          https://repo.continuum.io/pkgs/free/noarch
                          https://repo.continuum.io/pkgs/r/linux-64
                          https://repo.continuum.io/pkgs/r/noarch
                          https://repo.continuum.io/pkgs/pro/linux-64
                          https://repo.continuum.io/pkgs/pro/noarch
          package cache : /home/ondrej/anaconda/pkgs
                          /home/ondrej/.conda/pkgs
       envs directories : /home/ondrej/anaconda/envs
                          /home/ondrej/.conda/envs
               platform : linux-64
             user-agent : conda/4.4.10 requests/2.14.2 CPython/2.7.13 Linux/4.9.0-5-amd64 debian/9 glibc/2.24
                UID:GID : 1000:1000
             netrc file : None
           offline mode : False
```
</p></details>

<details open><summary><code>`conda config --show-sources`</code></summary><p>

```
==> /home/ondrej/.condarc <==
ssl_verify: True
channels:
  - anaconda-fusion
  - defaults
```
</p></details>

<details><summary><code>`conda list --show-channel-urls`</code></summary><p>

```
# packages in environment at /home/ondrej/anaconda:
#
# Name                    Version                   Build  Channel
_license                  1.1                      py27_1    defaults
_nb_ext_conf              0.3.0                    py27_0    defaults
abstract-rendering        0.5.1               np111py27_0    defaults
accelerate                2.0.2               np19py27_p0    defaults
accelerate_cudalib        2.0                           0    defaults
alabaster                 0.7.9                    py27_0    defaults
anaconda                  custom           py27h4a00acb_0    defaults
anaconda-clean            1.0.0                    py27_0    defaults
anaconda-client           1.6.0                    py27_0    defaults
anaconda-navigator        1.5.0                    py27_0    defaults
anaconda-project          0.4.1                    py27_0    defaults
apipkg                    1.4                       <pip>
argcomplete               1.0.0                    py27_1    defaults
astroid                   1.4.9                    py27_0    defaults
astropy                   1.3                 np111py27_0    defaults
babel                     2.3.4                    py27_0    defaults
backports                 1.0                      py27_0    defaults
backports_abc             0.5                      py27_0    defaults
bcolz                     0.11.3                   py27_0    defaults
beautiful-soup            4.3.2                    py27_0    <unknown>
beautifulsoup4            4.5.3                    py27_0    defaults
binstar                   0.12                          2    defaults
bitarray                  0.8.1                    py27_0    <unknown>
blaze                     0.10.1                   py27_0    defaults
blaze-core                0.9.0                    py27_0    defaults
blinker                   1.4                        py_0    conda-forge
bokeh                     0.12.4                   py27_0    defaults
boto                      2.45.0                   py27_0    defaults
bottleneck                1.2.0               np111py27_0    defaults
ca-certificates           2018.1.18                     0    conda-forge
cairo                     1.14.8                        0    defaults
cdecimal                  2.3                      py27_2    defaults
certifi                   2018.1.18                py27_0    defaults
cffi                      1.9.1                    py27_0    defaults
chardet                   2.3.0                     <pip>
chardet                   2.3.0                    py27_0    defaults
chest                     0.2.3                    py27_0    defaults
click                     6.7                      py27_0    defaults
cloudpickle               0.2.2                    py27_0    defaults
clyent                    1.2.2                    py27_0    defaults
colorama                  0.3.7                    py27_0    defaults
conda                     4.4.10                   py27_0    defaults
conda-build               1.19.0                   py27_0    defaults
conda-env                 2.6.0                h36134e3_1    defaults
conda-manager             0.3.1                    py27_0    defaults
configobj                 5.0.6                    py27_0    <unknown>
configparser              3.5.0                    py27_0    defaults
configparser              3.3.0.post2               <pip>
contextlib2               0.5.4                    py27_0    defaults
coverage                  4.0.3                     <pip>
cryptography              1.7.1                    py27_0    defaults
cudatoolkit               7.0                           1    defaults
curl                      7.52.1                        0    defaults
cycler                    0.10.0                   py27_0    defaults
cython                    0.25.2                   py27_0    defaults
cytoolz                   0.8.2                    py27_0    defaults
dask                      0.13.0                   py27_0    defaults
datashape                 0.5.4                    py27_0    defaults
dbus                      1.10.10                       0    defaults
decorator                 4.0.11                   py27_0    defaults
dill                      0.2.5                    py27_0    defaults
docutils                  0.13.1                   py27_0    defaults
doit                      0.29.0                    <pip>
dynd-python               0.7.2                    py27_0    defaults
ecdsa                     0.13                      <pip>
entrypoints               0.2.2                    py27_0    defaults
enum34                    1.1.6                    py27_0    defaults
et_xmlfile                1.0.1                    py27_0    defaults
execnet                   1.4.1                     <pip>
expat                     2.1.0                         0    defaults
fastcache                 1.0.2                    py27_1    defaults
fftw                      3.3.4                         0    https://conda.binstar.org/jakirkham
filelock                  2.0.6                    py27_0    defaults
flask                     0.12                     py27_0    defaults
flask-cors                3.0.2                    py27_0    defaults
flask-login               0.4.0                    py27_0    defaults
flask-sqlalchemy          2.3.2                      py_0    conda-forge
fontconfig                2.12.1                        2    defaults
freetype                  2.5.5                         2    defaults
funcsigs                  1.0.2                    py27_0    defaults
functools32               3.2.3.2                  py27_0    defaults
futures                   3.0.5                    py27_0    defaults
get_terminal_size         1.0.0                    py27_0    defaults
gevent                    1.2.1                    py27_0    defaults
glib                      2.50.2                        1    defaults
greenlet                  0.4.11                   py27_0    defaults
grin                      1.2.1                    py27_3    defaults
gst-plugins-base          1.8.0                         0    defaults
gstreamer                 1.8.0                         0    defaults
h5py                      2.6.0               np111py27_2    defaults
harfbuzz                  0.9.39                        2    defaults
hdf4                      4.2.11                        0    defaults
hdf5                      1.8.17                        1    defaults
heapdict                  1.0.0                    py27_1    defaults
hmmlearn                  0.2.0                     <pip>
hmmlearn                  0.3.0b              np111py27_0    omnia
holoviews                 1.8.3                    py27_0    defaults
icalendar                 3.10                      <pip>
icu                       54.1                          0    defaults
idna                      2.2                      py27_0    defaults
imagesize                 0.7.1                    py27_0    defaults
ipaddress                 1.0.18                   py27_0    defaults
ipdb                      0.8.1                     <pip>
ipykernel                 4.5.2                    py27_0    defaults
ipython                   5.1.0                    py27_0    defaults
ipython-notebook          4.0.4                    py27_0    defaults
ipython-qtconsole         4.0.1                    py27_0    defaults
ipython_genutils          0.1.0                    py27_0    defaults
ipywidgets                5.2.2                    py27_1    defaults
isort                     4.2.5                    py27_0    defaults
itsdangerous              0.24                     py27_0    <unknown>
javaobj-py3               0.2.3                     <pip>
jbig                      2.1                           0    defaults
jdcal                     1.3                      py27_0    defaults
jedi                      0.9.0                    py27_1    defaults
jinja2                    2.9.4                    py27_0    defaults
joblib                    0.8.4                    py27_0    defaults
jpeg                      9b                            0    defaults
jsonschema                2.5.1                    py27_0    defaults
jupyter                   1.0.0                    py27_3    defaults
jupyter_client            4.4.0                    py27_0    defaults
jupyter_console           5.0.0                    py27_0    defaults
jupyter_core              4.2.1                    py27_0    defaults
krb5                      1.13.2                        0    defaults
lazy-object-proxy         1.2.2                    py27_0    defaults
ldap3                     2.3              py27hcb53f9d_0    defaults
libdynd                   0.7.2                         0    defaults
libffi                    3.2.1                         1    defaults
libgcc                    4.8.5                         2    defaults
libgcc-ng                 7.2.0                h7cc24e2_2    defaults
libgfortran               3.0.0                         1    defaults
libiconv                  1.14                          0    defaults
libpng                    1.6.27                        0    defaults
libsodium                 1.0.10                        0    defaults
libtiff                   4.0.6                         3    defaults
libxcb                    1.12                          1    defaults
libxml2                   2.9.4                         0    defaults
libxslt                   1.1.29                        0    defaults
llvmlite                  0.15.0                   py27_0    defaults
locket                    0.2.0                    py27_1    defaults
lockfile                  0.12.2                    <pip>
logilab-common            1.0.2                    py27_0    defaults
luigi                     2.1.1                     <pip>
lxml                      3.7.2                    py27_0    defaults
markupsafe                0.23                     py27_2    defaults
matplotlib                2.0.0               np111py27_0    defaults
mistune                   0.7.3                    py27_0    defaults
mkl                       2017.0.1                      0    defaults
mkl-rt                    11.1                         p0    defaults
mkl-service               1.1.2                    py27_3    defaults
mklfft                    2.1                 np19py27_p0    defaults
mock                      2.0.0                    py27_0    defaults
mpmath                    0.19                     py27_1    defaults
multipledispatch          0.4.9                    py27_0    defaults
mysql                     5.5.24                        0    defaults
mysql-python              1.2.5                    py27_0    defaults
nb_anacondacloud          1.2.0                    py27_0    defaults
nb_conda                  2.0.0                    py27_0    defaults
nb_conda_kernels          2.0.0                    py27_0    defaults
nbconvert                 4.2.0                    py27_0    defaults
nbformat                  4.2.0                    py27_0    defaults
nbpresent                 3.0.2                    py27_0    defaults
networkx                  1.11                     py27_0    defaults
nltk                      3.2.2                    py27_0    defaults
nose                      1.3.7                    py27_1    defaults
notebook                  4.3.1                    py27_0    defaults
numba                     0.30.1              np111py27_0    defaults
numbapro                  0.23.1                  py27_p0    defaults
numbapro_cudalib          0.2                           0    defaults
numexpr                   2.6.1               np111py27_2    defaults
numpy                     1.11.3                   py27_0    defaults
numpydoc                  0.6.0                    py27_0    defaults
odo                       0.5.0                    py27_1    defaults
openldap                  2.4.36                        1    defaults
openpyxl                  2.4.1                    py27_0    defaults
openssl                   1.0.2n               hb7f436b_0    defaults
pandas                    0.19.2              np111py27_1    defaults
param                     1.5.1                    py27_0    defaults
paramiko                  1.16.0                    <pip>
partd                     0.3.7                    py27_0    defaults
patchelf                  0.9                           0    defaults
path.py                   10.0                     py27_0    defaults
pathlib2                  2.2.0                    py27_0    defaults
patsy                     0.4.1                    py27_0    defaults
pbr                       1.3.0                    py27_0    defaults
pcre                      8.39                          1    defaults
pep8                      1.7.0                    py27_0    defaults
pexpect                   4.2.1                    py27_0    defaults
pickleshare               0.7.4                    py27_0    defaults
pillow                    4.0.0                    py27_0    defaults
pip                       9.0.1                    py27_1    defaults
pixman                    0.34.0                        0    defaults
pkginfo                   1.3.2                    py27_0    defaults
ply                       3.9                      py27_0    defaults
progressbar               2.3                      py27_0    defaults
progressbar2              3.10.1                    <pip>
prompt_toolkit            1.0.9                    py27_0    defaults
psutil                    5.0.1                    py27_0    defaults
ptyprocess                0.5.1                    py27_0    defaults
py                        1.4.32                   py27_0    defaults
pyasn1                    0.1.9                    py27_0    defaults
pycairo                   1.10.0                   py27_0    defaults
pycosat                   0.6.3            py27ha4109ae_0    defaults
pycparser                 2.17                     py27_0    defaults
pycrypto                  2.6.1                    py27_4    defaults
pycurl                    7.43.0                   py27_2    defaults
pyfftw                    0.9.2                np19py27_1    https://conda.binstar.org/jakirkham
pyflakes                  1.5.0                    py27_0    defaults
pygments                  2.1.3                    py27_0    defaults
pyhdf                     0.9.0               np111py27_1    conda-forge
pyinotify                 0.9.6                     <pip>
pylint                    1.6.4                    py27_1    defaults
pyopenssl                 16.2.0                   py27_0    defaults
pyparsing                 2.1.4                    py27_0    defaults
pyqt                      5.6.0                    py27_2    defaults
pyserial                  2.7                      py27_0    defaults
pytables                  3.3.0               np111py27_0    defaults
pytest                    3.0.5                    py27_0    defaults
pytest-cov                2.2.1                     <pip>
python                    2.7.13                        0    defaults
python-daemon             2.1.1                     <pip>
python-dateutil           2.6.0                    py27_0    defaults
python-ldap               2.4.13                   py27_0    defaults
python-utils              2.0.0                     <pip>
pytz                      2016.10                  py27_0    defaults
pyyaml                    3.12                     py27_0    defaults
pyzmq                     16.0.2                   py27_0    defaults
qt                        5.6.2                         3    defaults
qtawesome                 0.4.3                    py27_0    defaults
qtconsole                 4.2.1                    py27_1    defaults
qtpy                      1.2.1                    py27_0    defaults
readline                  6.2                           2    defaults
redis                     3.2.0                         0    defaults
redis-py                  2.10.5                   py27_0    defaults
requests                  2.14.2                   py27_0    defaults
rk                        0.3b1                     <pip>
rope                      0.9.4                    py27_1    <unknown>
ruamel_yaml               0.11.14                  py27_1    defaults
ruffus                    2.6.3                     <pip>
runipy                    0.1.3                    py27_0    <unknown>
scandir                   1.4                      py27_0    defaults
scikit-image              0.12.3              np111py27_1    defaults
scikit-learn              0.18.1              np111py27_1    defaults
scipy                     0.18.1              np111py27_1    defaults
seaborn                   0.7.1                    py27_0    defaults
setuptools                38.4.0                   py27_0    defaults
simplegeneric             0.8.1                    py27_1    defaults
singledispatch            3.4.0.3                  py27_0    defaults
sip                       4.18                     py27_0    defaults
six                       1.10.0                   py27_0    defaults
snowballstemmer           1.2.1                    py27_0    defaults
sockjs-tornado            1.0.3                    py27_0    defaults
spectrum                  0.1                       <pip>
sphinx                    1.5.1                    py27_0    defaults
sphinx_rtd_theme          0.1.9                    py27_0    defaults
spyder                    3.1.2                    py27_0    defaults
sqlalchemy                1.1.5                    py27_0    defaults
sqlite                    3.13.0                        0    defaults
sqlsoup                   0.9.1                     <pip>
ssl_match_hostname        3.4.0.2                  py27_1    defaults
statsmodels               0.6.1               np111py27_1    defaults
subprocess32              3.2.7                    py27_0    defaults
sympy                     1.0                      py27_0    defaults
system                    5.8                           2    <unknown>
terminado                 0.6                      py27_0    defaults
theano                    0.5.0                np15py27_0    defaults
tk                        8.5.18                        0    defaults
toolz                     0.8.2                    py27_0    defaults
tornado                   4.4.2                    py27_0    defaults
traitlets                 4.3.1                    py27_0    defaults
ujson                     1.35                     py27_0    defaults
unicodecsv                0.14.1                   py27_0    defaults
util-linux                2.21                          0    <unknown>
wcwidth                   0.1.7                    py27_0    defaults
werkzeug                  0.11.15                  py27_0    defaults
wheel                     0.29.0                   py27_0    defaults
widgetsnbextension        1.2.6                    py27_0    defaults
wrapt                     1.10.8                   py27_0    defaults
xarray                    0.9.3                    py27_0    defaults
xlrd                      1.0.0                    py27_0    defaults
xlsxwriter                0.9.6                    py27_0    defaults
xlwt                      1.2.0                    py27_0    defaults
xmltodict                 0.10.2                    <pip>
xz                        5.2.2                         1    defaults
yaml                      0.1.6                         0    defaults
zeromq                    4.1.5                         0    defaults
zlib                      1.2.8                         3    defaults
```
</p></details>

### PR DESCRIPTION
**Pull request by mbargull**
See: https://github.com/conda/conda/issues/6739#issuecomment-362941051

### CODE REVIEW

**Comment by mbargull** [comment]
@kalefranz feel free to add a test ;)

File: conda/cli/main.py [hunk 1]
```diff
@@ -63,6 +63,34 @@ def init_loggers(context=None):
...
+def init_std_stream_encoding():
+    """
+    PY2 compat: Initialize encoding for standard streams.
+    Python 2 sets the encoding of stdout/stderr to None if not run in a
+    terminal context and thus falls back to ASCII.
+    """
+    for stream_name in ("stdout", "stderr"):
+        stream = getattr(sys, stream_name)
+        if not stream.encoding:
+            break
```
Comment (mbargull) [inline conda/cli/main.py]: This duplicate code is just to avoid the imports below if they are not neccessary.

**Comment by mbargull** [comment]
Oh, BTW, I noticed
```python
def _ensure_text_type(value):
    # copying here from conda/common/compat.py to avoid the import
```
. But imports from `conda.common.compat` should be nearly cost-free since that module is already imported in `conda.__init__` anyway, or not?

**Comment by kalefranz** [comment]
You’re right about the conda.common.compat import, but I think the import is wrong. I’d rather refactorvto use `.items()` instead of `iteritems()` and then I guess an in-line import for the rest.

**Comment by mbargull** [comment]
You'd need the name of the stream because of the `setattr`. But I think you are right, I like it much better if we don't `setattr` and make it:
```
sys.stdout = init_stream_encoding(sys.stdout)
sys.stderr = init_stream_encoding(sys.stderr)
```

**Comment by mbargull** [comment]
Have to rework the logic when the encoding is added though, as the tests show, e.g.,
```
    if not stream.encoding:
AttributeError: 'cStringIO.StringO' object has no attribute 'encoding'
```

**Comment by kalefranz** [comment]
I started playing around with some stuff here.  And pretty quickly ran into the same problems that the tests are exposing.  Haven't solved them yet.  I'll make a PR to this one though incase you want to incorporate those changes.

**Comment by kalefranz** [comment]
https://github.com/mbargull/conda/pull/1

**Comment by mbargull** [comment]
Also the Python 3 tests show
```
0.72s call     tests/core/test_solve.py::test_solve_1
=================================== FAILURES ===================================
_____________________________ TestJson.test_config _____________________________
Traceback (most recent call last):
  File "/home/circleci/conda/conda/exceptions.py", line 789, in __call__
    return func(*args, **kwargs)
  File "/home/circleci/conda/conda/cli/main.py", line 106, in _main
    exit_code = do_call(args, p)
  File "/home/circleci/conda/conda/cli/conda_argparse.py", line 77, in do_call
    exit_code = getattr(module, func_name)(args, parser)
  File "/home/circleci/conda/conda/cli/main_config.py", line 30, in execute
    execute_config(args, parser)
  File "/home/circleci/conda/conda/cli/main_config.py", line 336, in execute_config
    get=json_get
  File "/home/circleci/conda/conda/cli/common.py", line 170, in stdout_json_success
    stdout_json(result)
  File "/home/circleci/conda/conda/common/io.py", line 46, in decorated
    return f(*args, **kwds)
  File "/home/circleci/conda/conda/cli/common.py", line 163, in stdout_json
    json.dump(d, sys.stdout, indent=2, sort_keys=True, cls=EntityEncoder)
  File "/opt/conda/lib/python3.6/json/__init__.py", line 180, in dump
    fp.write(chunk)
  File "/opt/conda/lib/python3.6/codecs.py", line 377, in write
    self.stream.write(data)
TypeError: string argument expected, got 'bytes'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/circleci/conda/tests/test_cli.py", line 26, in test_config
    res = capture_json_with_argv('conda config --get --json')
  File "/home/circleci/conda/tests/helpers.py", line 67, in capture_json_with_argv
    stdout, stderr, exit_code = run_inprocess_conda_command(command)
  File "/home/circleci/conda/tests/helpers.py", line 100, in run_inprocess_conda_command
    exit_code = cli.main()
  File "/home/circleci/conda/conda/cli/main.py", line 155, in main
    return conda_exception_handler(_main, *args)
  File "/home/circleci/conda/conda/exceptions.py", line 1025, in conda_exception_handler
    return_value = exception_handler(func, *args, **kwargs)
  File "/home/circleci/conda/conda/exceptions.py", line 792, in __call__
    return self.handle_exception(exc_val, exc_tb)
  File "/home/circleci/conda/conda/exceptions.py", line 832, in handle_exception
    return self.handle_unexpected_exception(exc_val, exc_tb)
  File "/home/circleci/conda/conda/exceptions.py", line 844, in handle_unexpected_exception
    self.print_error_report(error_report)
  File "/home/circleci/conda/conda/exceptions.py", line 885, in print_error_report
    stdout_json(error_report)
  File "/home/circleci/conda/conda/common/io.py", line 46, in decorated
    return f(*args, **kwds)
  File "/home/circleci/conda/conda/cli/common.py", line 163, in stdout_json
    json.dump(d, sys.stdout, indent=2, sort_keys=True, cls=EntityEncoder)
  File "/opt/conda/lib/python3.6/json/__init__.py", line 180, in dump
    fp.write(chunk)
  File "/opt/conda/lib/python3.6/codecs.py", line 377, in write
    self.stream.write(data)
TypeError: string argument expected, got 'bytes'
______________________________ TestJson.test_list ______________________________
Traceback (most recent call last):
  File "/home/circleci/conda/conda/exceptions.py", line 789, in __call__
    return func(*args, **kwargs)
  File "/home/circleci/conda/conda/cli/main.py", line 106, in _main
    exit_code = do_call(args, p)
  File "/home/circleci/conda/conda/cli/conda_argparse.py", line 77, in do_call
    exit_code = getattr(module, func_name)(args, parser)
  File "/home/circleci/conda/conda/cli/main_list.py", line 155, in execute
    show_channel_urls=context.show_channel_urls)
  File "/home/circleci/conda/conda/cli/main_list.py", line 95, in print_packages
    stdout_json(output)
  File "/home/circleci/conda/conda/common/io.py", line 46, in decorated
    return f(*args, **kwds)
  File "/home/circleci/conda/conda/cli/common.py", line 163, in stdout_json
    json.dump(d, sys.stdout, indent=2, sort_keys=True, cls=EntityEncoder)
  File "/opt/conda/lib/python3.6/json/__init__.py", line 180, in dump
    fp.write(chunk)
  File "/opt/conda/lib/python3.6/codecs.py", line 377, in write
    self.stream.write(data)
TypeError: string argument expected, got 'bytes'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/circleci/conda/tests/test_cli.py", line 61, in test_list
    res = capture_json_with_argv('conda list --json')
  File "/home/circleci/conda/tests/helpers.py", line 67, in capture_json_with_argv
    stdout, stderr, exit_code = run_inprocess_conda_command(command)
  File "/home/circleci/conda/tests/helpers.py", line 100, in run_inprocess_conda_command
    exit_code = cli.main()
  File "/home/circleci/conda/conda/cli/main.py", line 155, in main
    return conda_exception_handler(_main, *args)
  File "/home/circleci/conda/conda/exceptions.py", line 1025, in conda_exception_handler
    return_value = exception_handler(func, *args, **kwargs)
  File "/home/circleci/conda/conda/exceptions.py", line 792, in __call__
    return self.handle_exception(exc_val, exc_tb)
  File "/home/circleci/conda/conda/exceptions.py", line 832, in handle_exception
    return self.handle_unexpected_exception(exc_val, exc_tb)
  File "/home/circleci/conda/conda/exceptions.py", line 844, in handle_unexpected_exception
    self.print_error_report(error_report)
  File "/home/circleci/conda/conda/exceptions.py", line 885, in print_error_report
    stdout_json(error_report)
  File "/home/circleci/conda/conda/common/io.py", line 46, in decorated
    return f(*args, **kwds)
  File "/home/circleci/conda/conda/cli/common.py", line 163, in stdout_json
    json.dump(d, sys.stdout, indent=2, sort_keys=True, cls=EntityEncoder)
  File "/opt/conda/lib/python3.6/json/__init__.py", line 180, in dump
    fp.write(chunk)
  File "/opt/conda/lib/python3.6/codecs.py", line 377, in write
    self.stream.write(data)
TypeError: string argument expected, got 'bytes'
```
which is interesting. Maybe this means some object outputs `bytes` via`auxlib.entity.EntityEncoder` in https://github.com/conda/conda/blob/4.4.10/conda/_vendor/auxlib/entity.py#L926-L939. Could be something to keep in mind.

**Comment by kalefranz** [comment]
Maybe https://github.com/conda/conda/blob/4.4.10/conda/_vendor/auxlib/entity.py#L528

```python
def box(self, instance, instance_type, val):
    return text_type(val) if isinstance(val, NumberField._type) else val
```

should just be

```python
def box(self, instance, instance_type, val):
    return text_type(val)
```

**Comment by kalefranz** [comment]
> `TypeError: string argument expected, got 'bytes'`

😫

**Comment by mbargull** [comment]
Yeah, I'm backing off and want to think about corner cases before applying the changes to the standard streams.

**Comment by mbargull** [comment]
Is the 4.4.11 merge window closed or still open? Should we sneak this one with the more "surgical" `PS1` fix into that release or rather the next?

**Comment by kalefranz** [comment]
I just tagged 4.4.11.  I guess I should have asked if you were close here or not.  We can bump the tag I guess.

**Comment by kalefranz** [comment]
I just removed the 4.4.11 tag.  What's the surgical PS1 fix you're talking about?  I missed that in the conversation somewhere I guess.  Seems like there are a lot of issues flying around this week!  (And I was offline Sat-Tues, which doesn't help.)

**Comment by mbargull** [comment]
That commit: https://github.com/conda/conda/pull/6909/commits/3c4c84763c7c903d91f407a563226de787221ca9

**Comment by mbargull** [comment]
I guess the main problem people had was with Unicode characters in `PS1` which Python 2 choked on outputting without an IO encoding set. With `ensure_binary` the `print` doesn't have to use the stream's encoding then.

**Comment by mbargull** [comment]
Well that was a bad timing, I hadn't intended to delay the 4.4.11 release because these changes (IMHO) aren't the most important ones.
Well, apart from for your perspective it should be nice to have them included, so that fewer Python 2 users bug you with their `ValueError: unsupported format character '{'` errors :wink:

**Pull request merged**